gk/main.c

```diff
@@ -968,12 +972,212 @@ gk_setup_rss(struct gk_config *gk_conf)
+xmit_icmpv6(struct rte_mbuf *pkt, uint16_t *num_pkts,
+	struct rte_mbuf **icmp_bufs)
+{
+	uint8_t ip_addr_tmp[16];
+	struct ether_addr eth_addr_tmp;
+	struct ether_hdr *icmp_eth;
+	struct ipv6_hdr *icmp_ipv6;
+	struct icmpv6_hdr *icmpv6_hdr;
+	int icmpv6_pkt_len = sizeof(struct ether_hdr) +
+		sizeof(struct ipv6_hdr) + sizeof(struct icmpv6_hdr);
+	if (pkt->data_len > icmpv6_pkt_len) {
+		int ret = rte_pktmbuf_trim(pkt,
+			pkt->data_len - icmpv6_pkt_len);
+		if (ret < 0) {
+			RTE_LOG(ERR, MBUF,
+				"Failed to remove %d bytes of data at the end of the mbuf at %s.",
+				pkt->data_len - icmpv6_pkt_len, __func__);
+			return;
+		}
+
+		icmp_eth = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	} else {
+		icmp_eth = (struct ether_hdr *)rte_pktmbuf_append(pkt,
+			icmpv6_pkt_len - pkt->data_len);
+		if (icmp_eth == NULL) {
+			RTE_LOG(ERR, MBUF,
+				"Failed to append %d bytes of new data: not enough headroom space in the first segment at %s!\n",
+				icmpv6_pkt_len - pkt->data_len, __func__);
+			return;
+		}
+	}
+
+	eth_addr_tmp = icmp_eth->s_addr;
+	ether_addr_copy(&icmp_eth->d_addr, &icmp_eth->s_addr);
+	ether_addr_copy(&eth_addr_tmp, &icmp_eth->d_addr);
+
+	/* Set-up IPv6 header. */
+	icmp_ipv6 = (struct ipv6_hdr *)&icmp_eth[1];
+	icmp_ipv6->vtc_flow = rte_cpu_to_be_32(IPv6_DEFAULT_VTC_FLOW);
+	icmp_ipv6->payload_len = rte_cpu_to_be_16(sizeof(*icmpv6_hdr));
+	icmp_ipv6->proto = IPPROTO_ICMPV6;
+	/*
+	 * The IP Hop Limit field must be 255 as required by
+	 * RFC 4861, sections 7.1.1 and 7.1.2.
+	 */
+	icmp_ipv6->hop_limits = 255;
+	rte_memcpy(ip_addr_tmp, icmp_ipv6->src_addr, sizeof(ip_addr_tmp));
+	rte_memcpy(icmp_ipv6->src_addr, icmp_ipv6->dst_addr,
+		sizeof(icmp_ipv6->src_addr));
+	rte_memcpy(icmp_ipv6->dst_addr, ip_addr_tmp,
+		sizeof(icmp_ipv6->dst_addr));
+
+	/* Set-up ICMPv6 header. */
+	icmpv6_hdr = (struct icmpv6_hdr *)&icmp_ipv6[1];
+	icmpv6_hdr->type = ICMPV6_TIME_EXCEED;
+	icmpv6_hdr->code = ICMPV6_EXC_HOPLIMIT;
+	icmpv6_hdr->cksum = 0; /* Calculated below. */
+
+	icmpv6_hdr->cksum = rte_ipv6_icmpv6_cksum(icmp_ipv6, icmpv6_hdr);
+
+	icmp_bufs[*num_pkts] = pkt;
+	(*num_pkts)++;
+}
+
+/*
+ * For IPv4, according to the RFC 1812 section 5.3.1 Time to Live (TTL),
+ * if the TTL is reduced to zero (or less), the packet MUST be
+ * discarded, and if the destination is not a multicast address the
+ * router MUST send an ICMP Time Exceeded message, Code 0 (TTL Exceeded
+ * in Transit) message to the source.
+ *
+ * For IPv6, according to the RFC 1883 section 4.4,
+ * if the IPv6 Hop Limit is less than or equal to 1, then the router needs to
+ * send an ICMP Time Exceeded -- Hop Limit Exceeded in Transit message to
+ * the Source Address and discard the packet.
+ */
+static int
+update_ip_hop_count(struct ipacket *packet, uint16_t *num_pkts,
+	struct rte_mbuf **icmp_bufs)
+{
+	if (packet->flow.proto == ETHER_TYPE_IPv4) {
+		struct ipv4_hdr *ipv4_hdr =
+			rte_pktmbuf_mtod_offset(packet->pkt,
+			struct ipv4_hdr *, sizeof(struct ether_hdr));
```

Comment:
Ethernet header has variable size, use the information in `packet` to avoid parsing everything again.

Analogous issue and solution in the IPv6 case.